mtags-java/src/main/scala/scala/meta/internal/pc/JavaHoverProvider.scala

```diff
@@ -0,0 +1,280 @@
+      if (element.getKind == CONSTRUCTOR)
+        element.getEnclosingElement.getSimpleName
+      else element.getSimpleName
+    val arguments =
+      element.getParameters.asScala.map(argumentHover).mkString(", ")
+
+    val throws = element.getThrownTypes.asScala
+    val throwsHover =
+      if (throws.isEmpty) ""
+      else
+        throws
+          .map(t => t.accept(new JavaTypeVisitor(), null))
+          .mkString(" throws ", ", ", "")
+
+    s"$modifiers$returnType $functionName($arguments)$throwsHover"
+  }
+
+  private def packageHover(element: PackageElement): String =
+    s"package ${element.getQualifiedName}"
+
+  private def variableHover(element: VariableElement): String = {
+    val modifiers = modifiersHover(element)
+    val variableType = typeHover(element.asType())
+    val name = element.getSimpleName
+
+    s"$modifiers$variableType $name"
+  }
+
+  private def documentation(element: Element): String = {
+    element match {
+      case symbol: Symbol =>
+        val sym = semanticdbSymbol(symbol)
+
+        compiler.search
+          .documentation(
+            sym,
+            new ParentSymbols {
+              override def parents(): util.List[String] = symbol.owner
+                .asType() match {
+                case cType: ClassType => overriddenSymbols(symbol, cType)
+                case _ => util.Collections.emptyList[String]
+              }
+            },
+          )
+          .toScala
+          .map(_.docstring())
+          .getOrElse("")
+      case _ => ""
+    }
+  }
+
+  private def overriddenSymbols(
+      symbol: Symbol,
+      cType: ClassType,
+  ): util.List[String] = {
+    val types = baseSymbols(cType)
+
+    (for {
+      t <- types
+      s <- t.tsym.getEnclosedElements.asScala
+      if s.getSimpleName == symbol.getSimpleName
+    } yield semanticdbSymbol(s)).asJava
+  }
+
+  private def baseSymbols(cType: ClassType): List[ClassType] = {
+    val superType = Option(cType.supertype_field)
+    val inheritedTypes =
+      if (cType.interfaces_field == null) List()
+      else cType.interfaces_field.asScala
+
+    val baseTypes = (superType ++ inheritedTypes).collect { case c: ClassType =>
+      c
+    }
+
+    baseTypes.flatMap(c => c :: baseSymbols(c)).toList
+  }
+
+  private def semanticdbSymbol(symbol: Symbol): String = {
+
+    def descriptors(symbol: Symbol): List[Descriptor] = {
+      if (symbol == null || symbol.name.toString == "") Empty :: Nil
+      else {
+        val owner = descriptors(symbol.owner)
+        val desc = {
+          val name = symbol.name.toString
+
+          symbol match {
+            case _: PackageSymbol => Package(name)
+            case m: MethodSymbol => Method(name, disambiguator(m))
+            case _: ClassSymbol => Class(name)
+            case _: TypeVariableSymbol => TypeVariable(name)
+            case _: VarSymbol => Var(name)
+            case _ => Empty
+          }
+        }
+
+        desc :: owner
+      }
+    }
```

Comment:
Minor suggestioin - this might be converted into @tailrec variant:
```scala
@tailrec
def descriptors(acc: List[Descriptor], sym: Symbol): List[Descriptor] = {
  if (symbol == null || symbol.name.toString == "") {
    if (acc.isEmpty) Empty :: Nil
    else acc
  } else  {
    val desc = ...
    descriptors(desc :: acc, symbol.owner)
  }
}
```